File: ArchiMeter.Analysis.FS/XamlCodeWriter.fs
namespace ArchiMeter.Analysis.Xaml
open System
open System.Text.RegularExpressions
open Roslyn.Compilers.Common
open Roslyn.Compilers.CSharp
open Roslyn.Services
open Roslyn.Services.Formatting
open ArchiMeter.Common.Xaml

type internal XamlCodeWriter() =
    static let BindingRegex = new Regex("^{Binding(.+)?}$", RegexOptions.Compiled)

    let getPropertyIdentifier(prop : XamlPropertyNode) =
        match prop.Value with 
        | :? XamlNode as xn -> Syntax.IdentifierName(xn.VariableName) 
        | null -> Syntax.IdentifierName(Syntax.Token(SyntaxKind.NullKeyword)) 
        | _ as o -> Syntax.IdentifierName(o.ToString())

    let createBindingSyntax(nodeClassName : string, nodeVariableName : string, variableName : string, inlineValue : string) =
        let strings = inlineValue.Split(',') |> Seq.toArray
        let parameters = strings 
                         |> Seq.collect (fun p -> 
                                              p.Trim().Split([|','|], StringSplitOptions.RemoveEmptyEntries)
                                              |> Seq.map (fun kv -> kv.Split('='))
                                              |> Seq.map (fun t -> (t |> Seq.head, String.Join("=", t |> Seq.skip 1))))
                         |> Seq.map (fun (k,v) -> Syntax.BinaryExpression(SyntaxKind.AssignExpression, Syntax.IdentifierName(k), Syntax.IdentifierName(v.Replace("'","\""))))
        let separators = Seq.init (strings.Length - 1) (fun x -> Syntax.Token(SyntaxKind.CommaToken))
        let binding = Syntax.ObjectCreationExpression(
                                                      Syntax.ParseTypeName("Binding"), 
                                                      null, 
                                                      Syntax.InitializerExpression(
                                                                                    SyntaxKind.ObjectInitializerExpression,
                                                                                    Syntax.SeparatedList<ExpressionSyntax>(parameters |> Seq.cast<ExpressionSyntax>, separators)))
        let memberAccess = Syntax.MemberAccessExpression(SyntaxKind.MemberAccessExpression, Syntax.IdentifierName(nodeVariableName), Syntax.IdentifierName("SetBinding"))
        let dependencyProperty = Syntax.Argument(Syntax.MemberAccessExpression(SyntaxKind.MemberAccessExpression, Syntax.IdentifierName(nodeClassName), Syntax.IdentifierName(variableName + "Property")))
        let arguments = [|dependencyProperty; Syntax.Argument(binding)|]
        let memberInvocation = Syntax.InvocationExpression(memberAccess, Syntax.ArgumentList(Syntax.SeparatedList(arguments, [|Syntax.Token(SyntaxKind.CommaToken)|])))
        Syntax.ExpressionStatement(memberInvocation)

    let createDependencyPropertySyntax(variableName : string) (attr : XamlNodeAttribute) =
        let memberAccess = Syntax.MemberAccessExpression(SyntaxKind.MemberAccessExpression, Syntax.IdentifierName(variableName), Syntax.IdentifierName("SetValue"))
        let argList = Syntax.SeparatedList([|Syntax.Argument(Syntax.IdentifierName(attr.VariableName + "Property")); Syntax.Argument(Syntax.IdentifierName(attr.Value))|], [|Syntax.Token(SyntaxKind.CommaToken)|])
        let invocation = Syntax.InvocationExpression(memberAccess, Syntax.ArgumentList(argList))
        Syntax.ExpressionStatement(invocation)

    let createLocalDeclaration(childNode : XamlNode) = 
        Syntax.LocalDeclarationStatement(    
            Syntax.VariableDeclaration(
                Syntax.IdentifierName("var"),
                    Syntax.SeparatedList(
                        Syntax.VariableDeclarator(
                            Syntax.Identifier(childNode.VariableName),
                            null,
                            Syntax.EqualsValueClause(Syntax.ObjectCreationExpression(Syntax.ParseTypeName(childNode.ClassName), Syntax.ArgumentList(), null))))))

    let resolveKnownPropertyName(name : string) =
        match name with
        | "Resources" -> "ResourceDictionary"
        | _ -> name

    let createParentSyntax(childNode : XamlNode) =
        let parent = childNode.Parent
        match parent.ClassName with
        | "Canvas"
        | "StackPanel"
        | "DockPanel"
        | "Grid" ->
                    let childrenAccess = Syntax.MemberAccessExpression(SyntaxKind.MemberAccessExpression, Syntax.IdentifierName(parent.VariableName), Syntax.IdentifierName("Children"))
                    let addAccess = Syntax.MemberAccessExpression(SyntaxKind.MemberAccessExpression, childrenAccess, Syntax.IdentifierName("Add"))
                    Syntax.ExpressionStatement(Syntax.InvocationExpression(addAccess, Syntax.ArgumentList(Syntax.SeparatedList(Syntax.Argument(Syntax.IdentifierName(childNode.VariableName))))))
        | "Border" ->
                      let childAccess = Syntax.MemberAccessExpression(SyntaxKind.MemberAccessExpression, Syntax.IdentifierName(parent.VariableName), Syntax.IdentifierName("Child"))
                      Syntax.ExpressionStatement(Syntax.BinaryExpression(SyntaxKind.AssignExpression, childAccess, Syntax.Token(SyntaxKind.EqualsToken), Syntax.IdentifierName(childNode.VariableName)))
        | "Window"
        | "UserControl" ->
                           let contentAccess = Syntax.MemberAccessExpression(SyntaxKind.MemberAccessExpression, Syntax.IdentifierName(parent.VariableName), Syntax.IdentifierName("Content"))
                           Syntax.ExpressionStatement(Syntax.BinaryExpression(SyntaxKind.AssignExpression, contentAccess, Syntax.Token(SyntaxKind.EqualsToken), Syntax.IdentifierName(childNode.VariableName)))
        | "ResourceDictionary" ->
                                  let addAccess = Syntax.MemberAccessExpression(SyntaxKind.MemberAccessExpression, Syntax.IdentifierName(parent.VariableName), Syntax.IdentifierName("Add"))
                                  let keyAttribute = childNode.Attributes |> Seq.find (fun a -> a.VariableName = "Key")
                                  let arguments = [|Syntax.Argument(Syntax.IdentifierName(keyAttribute.Value)); Syntax.Argument(Syntax.IdentifierName(childNode.VariableName))|]
                                  let separators = [|Syntax.Token(SyntaxKind.CommaToken)|]
                                  Syntax.ExpressionStatement(Syntax.InvocationExpression(addAccess, Syntax.ArgumentList(Syntax.SeparatedList(arguments, separators))))
        | _ ->
               let variableAccess = Syntax.MemberAccessExpression(SyntaxKind.MemberAccessExpression, Syntax.IdentifierName(parent.VariableName), Syntax.IdentifierName("AddChild"))
               Syntax.ExpressionStatement(Syntax.InvocationExpression(variableAccess, Syntax.ArgumentList(Syntax.SeparatedList(Syntax.Argument(Syntax.IdentifierName(childNode.VariableName))))))

    let isPropertyOf(parent : XamlNode, child : XamlNode) =
        let p = parent.Properties
        p |> Seq.exists (fun n -> n.Name = child.VariableName)

    let writeProperties (propertyCreation : XamlPropertyNode -> seq<StatementSyntax>) (childNode : XamlNode) =
        seq {
            for prop in childNode.Properties do
                yield! propertyCreation prop
                let memberAccess = Syntax.MemberAccessExpression(SyntaxKind.MemberAccessExpression, Syntax.IdentifierName(childNode.VariableName), Syntax.IdentifierName(prop.Name))
                let setAccess = Syntax.ExpressionStatement(Syntax.BinaryExpression(SyntaxKind.AssignExpression, memberAccess, Syntax.Token(SyntaxKind.EqualsToken), getPropertyIdentifier(prop)))
                yield setAccess :> StatementSyntax
        }

    let writeAttributes (childNode : XamlNode) =
        seq {
            for attr in childNode.Attributes do
                    if not (String.IsNullOrWhiteSpace attr.ExtraCode) then
                        yield Syntax.ExpressionStatement(Syntax.ParseExpression(attr.ExtraCode)) :> StatementSyntax
                    let bindingMatch = BindingRegex.Match attr.Value
                    if bindingMatch.Success then
                        let inlineValue = bindingMatch.Groups.[1].Value.Trim()
                        yield createBindingSyntax(childNode.ClassName, childNode.VariableName, attr.VariableName, inlineValue) :> StatementSyntax
                    elif attr.VariableName.Contains(".") then
                        yield createDependencyPropertySyntax childNode.VariableName attr :> StatementSyntax
                    else
                        let memberAccess = Syntax.MemberAccessExpression(SyntaxKind.MemberAccessExpression, Syntax.IdentifierName(childNode.VariableName), Syntax.IdentifierName(attr.VariableName))
                        yield Syntax.ExpressionStatement(
                                Syntax.BinaryExpression(
                                    SyntaxKind.AssignExpression, memberAccess, Syntax.Token(SyntaxKind.EqualsToken), Syntax.IdentifierName(attr.Value))) :> StatementSyntax
        }

    let rec generateChildStatement(childNode : XamlNode, propertyCreation : XamlPropertyNode -> seq<StatementSyntax>) =
        seq{
            yield createLocalDeclaration childNode :> StatementSyntax
            yield! writeProperties propertyCreation childNode
            yield! writeAttributes childNode
            if childNode.Parent <> null && not (isPropertyOf(childNode.Parent, childNode)) then 
                yield createParentSyntax(childNode) :> StatementSyntax
            yield! childNode.Children |> Seq.collect (fun c -> generateChildStatement(c, propertyCreation))
        }

    let createPropertyAssignment(prop : XamlPropertyNode, ownerExpression : ExpressionSyntax) = 
        match prop.Value with
                | :? seq<XamlNode> as values -> values |> Seq.map (fun v ->
                                                                            let memberAccess = Syntax.MemberAccessExpression(SyntaxKind.MemberAccessExpression, ownerExpression, Syntax.IdentifierName(prop.Name))
                                                                            let addAccess = Syntax.MemberAccessExpression(SyntaxKind.MemberAccessExpression, memberAccess, Syntax.IdentifierName("Add"))
                                                                            Syntax.ExpressionStatement(Syntax.InvocationExpression(addAccess, Syntax.ArgumentList(Syntax.SeparatedList(Syntax.Argument(Syntax.IdentifierName(v.VariableName)))))) :> StatementSyntax)
                                                        |> Seq.toArray
                                                        |> Seq.ofArray
                | _ -> seq
                          {
                            let memberAccess = Syntax.MemberAccessExpression(SyntaxKind.MemberAccessExpression, ownerExpression, Syntax.IdentifierName(prop.Name))
                            yield Syntax.ExpressionStatement(Syntax.BinaryExpression(SyntaxKind.AssignExpression, memberAccess, Syntax.Token(SyntaxKind.EqualsToken), getPropertyIdentifier prop)) :> StatementSyntax
                          }

    let rec createPropertySyntax(property : XamlPropertyNode, ownerSyntax : ExpressionSyntax) =
        match property.Value with
        | :? XamlNode as node -> generateChildStatement(node, (fun p -> createPropertySyntax(p, ownerSyntax)))
        | :? seq<XamlNode> as nodes -> nodes |> Seq.collect (fun n -> generateChildStatement(n, (fun p -> createPropertySyntax(p, ownerSyntax))))
        | _ -> Seq.empty

    let createSyntax(nodes : seq<XamlNode>, ownerSyntax : ExpressionSyntax) =
        nodes 
        |> Seq.collect (fun n -> generateChildStatement(n, fun p -> createPropertySyntax(p, ownerSyntax)))
        |> Seq.map (fun s -> SyntaxExtensions.WithTrailingTrivia(s, Syntax.ElasticCarriageReturnLineFeed))

    member this.CreateSyntax(node : XamlNode) =
        let properties = node.Properties |> Seq.collect (fun p -> createPropertySyntax(p, Syntax.ThisExpression()))
        let assignments = node.Properties |> Seq.collect (fun p -> createPropertyAssignment(p, Syntax.ThisExpression()))

        let codeRoot = Seq.concat [|properties; assignments; createSyntax (node.Children, Syntax.IdentifierName(node.VariableName))|]
        let classDeclarationSyntax = match String.IsNullOrWhiteSpace(node.BaseClassName) with 
                                     | false -> Syntax.ClassDeclaration(node.ClassName)
                                                      .AddModifiers(Syntax.Token(SyntaxKind.PublicKeyword))
                                                      .AddModifiers(Syntax.Token(SyntaxKind.PartialKeyword))
                                     | _ ->  Syntax.ClassDeclaration(node.ClassName)
                                                   .AddModifiers(Syntax.Token(SyntaxKind.PublicKeyword))
                                                   .AddModifiers(Syntax.Token(SyntaxKind.PartialKeyword))
                                                   .WithBaseList(Syntax.BaseList(
                                                                            Syntax.SeparatedList(
                                                                                Syntax.ParseTypeName(node.BaseClassName))))
        let xamlClass = classDeclarationSyntax
                            .WithMembers(
                                Syntax.List<MemberDeclarationSyntax>(
                                    Syntax.MethodDeclaration(
                                        Syntax.PredefinedType(
                                            Syntax.Token(SyntaxKind.VoidKeyword)),
                                        "InitializeComponent")
                                        .WithModifiers(
                                            Syntax.TokenList(Syntax.Token(SyntaxKind.ProtectedKeyword)))
                                        .WithBody(Syntax.Block(codeRoot))))
        let usings = Syntax.List(node.Usings |> Seq.map (fun kvp -> if String.IsNullOrWhiteSpace(kvp.Value) then
                                                                        Syntax.UsingDirective(Syntax.ParseName(kvp.Key)) 
                                                                       else 
                                                                        Syntax.UsingDirective(Syntax.NameEquals(Syntax.IdentifierName(kvp.Value)), Syntax.ParseName(kvp.Key))))
        let namespaceName = match String.IsNullOrWhiteSpace(node.NamespaceName) with | true -> "ArchiMeter" | _ -> node.NamespaceName
        let namespaceSyntax = Syntax.NamespaceDeclaration(Syntax.ParseName(namespaceName)).WithUsings(usings).AddMembers(xamlClass).Format(FormattingOptions.GetDefaultOptions()).GetFormattedRoot() :?> MemberDeclarationSyntax
        let compilationUnit = Syntax.CompilationUnit().AddMembers(namespaceSyntax)
        SyntaxTree.Create(compilationUnit)

File: ArchiMetrics.Analysis/XamlCodeWriter.cs
// --------------------------------------------------------------------------------------------------------------------
// <copyright file="XamlCodeWriter.cs" company="Reimers.dk">
//   Copyright © Reimers.dk 2013
//   This source is subject to the Microsoft Public License (Ms-PL).
//   Please see http://go.microsoft.com/fwlink/?LinkID=131993 for details.
//   All other rights reserved.
// </copyright>
// <summary>
//   Defines the XamlCodeWriter type.
// </summary>
// --------------------------------------------------------------------------------------------------------------------

namespace ArchiMetrics.Analysis
{
	using System;
	using System.Collections.Generic;
	using System.Linq;
	using System.Text.RegularExpressions;
	using ArchiMetrics.Common.Xaml;
	using Microsoft.CodeAnalysis;
	using Microsoft.CodeAnalysis.CSharp;
	using Microsoft.CodeAnalysis.CSharp.Syntax;
	using Microsoft.CodeAnalysis.Formatting;

	internal class XamlCodeWriter
	{
		private static readonly Regex BindingRegex = new Regex("^{Binding(.+)?}$", RegexOptions.Compiled);

		public SyntaxTree CreateSyntax(XamlNode node)
		{
			var properties = node.Properties.SelectMany(CreatePropertySyntax);
			var assignments = node.Properties.SelectMany(p => CreatePropertyAssignment(p, SyntaxFactory.ThisExpression()));

			var codeRoot = properties.Concat(assignments).Concat(CreateSyntax(node.Children));
			var classDeclarationSyntax =
				!string.IsNullOrWhiteSpace(node.BaseClassName)
					? SyntaxFactory.ClassDeclaration(node.ClassName)
							.AddModifiers(SyntaxFactory.Token(SyntaxKind.PublicKeyword))
							.AddModifiers(SyntaxFactory.Token(SyntaxKind.PartialKeyword))
					: SyntaxFactory.ClassDeclaration(node.ClassName)
							.AddModifiers(SyntaxFactory.Token(SyntaxKind.PublicKeyword))
							.AddModifiers(SyntaxFactory.Token(SyntaxKind.PartialKeyword))
							.WithBaseList(SyntaxFactory.BaseList(
								SyntaxFactory.SeparatedList(
									new[] { SyntaxFactory.ParseTypeName(node.BaseClassName) })));

			var xamlClass = classDeclarationSyntax
				.WithMembers(
					SyntaxFactory.List<MemberDeclarationSyntax>(
						new[]
						{
							SyntaxFactory.MethodDeclaration(
								SyntaxFactory.PredefinedType(
									SyntaxFactory.Token(SyntaxKind.VoidKeyword)),
								"InitializeComponent")
								.WithModifiers(
									SyntaxFactory.TokenList(SyntaxFactory.Token(SyntaxKind.ProtectedKeyword)))
								.WithBody(SyntaxFactory.Block(codeRoot))
						}));
			var usings = SyntaxFactory.List(node.Usings.Select(kvp => string.IsNullOrWhiteSpace(kvp.Value)
																   ? SyntaxFactory.UsingDirective(SyntaxFactory.ParseName(kvp.Key))
																   : SyntaxFactory.UsingDirective(SyntaxFactory.NameEquals(SyntaxFactory.IdentifierName(kvp.Value)), SyntaxFactory.ParseName(kvp.Key))));
			var namespaceName = string.IsNullOrWhiteSpace(node.NamespaceName) ? "ArchiMetrics" : node.NamespaceName;
			var namespaceSyntax =
				SyntaxFactory.NamespaceDeclaration(SyntaxFactory.ParseName(namespaceName))
					.WithUsings(usings)
					.AddMembers(xamlClass);
			//.Format(FormattingOptions.SmartIndent)
			//.GetFormattedRoot() as MemberDeclarationSyntax;
			var compilationUnit = SyntaxFactory.CompilationUnit().AddMembers(namespaceSyntax);
			return CSharpSyntaxTree.Create(compilationUnit);
		}

		private static ExpressionStatementSyntax CreateDependencyPropertySyntax(string variableName, XamlNodeAttribute attr)
		{
			var memberAccess = SyntaxFactory.MemberAccessExpression(SyntaxKind.MemberAccessExpression, SyntaxFactory.IdentifierName(variableName), SyntaxFactory.IdentifierName("SetValue"));
			var argList = SyntaxFactory.SeparatedList(
				new[] { SyntaxFactory.Argument(SyntaxFactory.IdentifierName(attr.VariableName + "Property")), SyntaxFactory.Argument(SyntaxFactory.IdentifierName(attr.Value)) },
				new[] { SyntaxFactory.Token(SyntaxKind.CommaToken) });
			var invocation = SyntaxFactory.InvocationExpression(memberAccess, SyntaxFactory.ArgumentList(argList));
			return SyntaxFactory.ExpressionStatement(invocation);
		}

		private static LocalDeclarationStatementSyntax CreateLocalDeclaration(XamlNode childNode)
		{
			var variables = SyntaxFactory.SeparatedList(
				new[]
				{
					SyntaxFactory.VariableDeclarator(
						SyntaxFactory.Identifier(childNode.VariableName),
						null,
						SyntaxFactory.EqualsValueClause(
							SyntaxFactory.ObjectCreationExpression(SyntaxFactory.ParseTypeName(childNode.ClassName), SyntaxFactory.ArgumentList(), null)))
				});
			return SyntaxFactory.LocalDeclarationStatement(
				SyntaxFactory.VariableDeclaration(
					SyntaxFactory.IdentifierName("var"),
					variables));
		}

		private static ExpressionStatementSyntax CreateParentSyntax(XamlNode childNode)
		{
			var parent = childNode.Parent;
			switch (parent.ClassName)
			{
				case "Canvas":
				case "StackPanel":
				case "DockPanel":
				case "Grid":
					{
						var childrenAccess = SyntaxFactory.MemberAccessExpression(SyntaxKind.MemberAccessExpression, SyntaxFactory.IdentifierName(parent.VariableName), SyntaxFactory.IdentifierName("Children"));
						var addAccess = SyntaxFactory.MemberAccessExpression(SyntaxKind.MemberAccessExpression, childrenAccess, SyntaxFactory.IdentifierName("Add"));
						return SyntaxFactory.ExpressionStatement(SyntaxFactory.InvocationExpression(addAccess, SyntaxFactory.ArgumentList(SyntaxFactory.SeparatedList(new[] { SyntaxFactory.Argument(SyntaxFactory.IdentifierName(childNode.VariableName)) }))));
					}

				case "Border":
					{
						var childAccess = SyntaxFactory.MemberAccessExpression(SyntaxKind.MemberAccessExpression, SyntaxFactory.IdentifierName(parent.VariableName), SyntaxFactory.IdentifierName("Child"));
						return SyntaxFactory.ExpressionStatement(SyntaxFactory.BinaryExpression(
							SyntaxKind.AssignExpression,
							childAccess,
							SyntaxFactory.Token(SyntaxKind.EqualsToken),
							SyntaxFactory.IdentifierName(childNode.VariableName)));
					}

				case "Window":
				case "UserControl":
					{
						var contentAccess = SyntaxFactory.MemberAccessExpression(
							SyntaxKind.MemberAccessExpression,
							SyntaxFactory.IdentifierName(parent.VariableName),
							SyntaxFactory.IdentifierName("Content"));
						return SyntaxFactory.ExpressionStatement(
							SyntaxFactory.BinaryExpression(
								SyntaxKind.AssignExpression,
								contentAccess,
								SyntaxFactory.Token(SyntaxKind.EqualsToken),
								SyntaxFactory.IdentifierName(childNode.VariableName)));
					}

				case "ResourceDictionary":
					{
						var
							addAccess = SyntaxFactory.MemberAccessExpression(SyntaxKind.MemberAccessExpression, SyntaxFactory.IdentifierName(parent.VariableName), SyntaxFactory.IdentifierName("Add"));
						var keyAttribute = childNode.Attributes.First(a => a.VariableName == "Key");
						var arguments = new[]
							                {
								                SyntaxFactory.Argument(SyntaxFactory.IdentifierName(keyAttribute.Value)), 
												SyntaxFactory.Argument(SyntaxFactory.IdentifierName(childNode.VariableName))
							                };
						var separators = new[] { SyntaxFactory.Token(SyntaxKind.CommaToken) };
						return SyntaxFactory.ExpressionStatement(SyntaxFactory.InvocationExpression(addAccess, SyntaxFactory.ArgumentList(SyntaxFactory.SeparatedList(arguments, separators))));
					}

				default:
					var variableAccess = SyntaxFactory.MemberAccessExpression(SyntaxKind.MemberAccessExpression, SyntaxFactory.IdentifierName(parent.VariableName), SyntaxFactory.IdentifierName("AddChild"));
					return
						SyntaxFactory.ExpressionStatement(
							SyntaxFactory.InvocationExpression(
								variableAccess,
								SyntaxFactory.ArgumentList(
									SyntaxFactory.SeparatedList(
										new[]
										{
											SyntaxFactory.Argument(SyntaxFactory.IdentifierName(childNode.VariableName))
										}))));
			}
		}

		private static bool IsPropertyOf(XamlNode parent, XamlNode child)
		{
			var p = parent.Properties;
			return p.Any(n => n.Name == child.VariableName);
		}

		private static StatementSyntax CreateBindingSyntax(
			string nodeClassName,
			string nodeVariableName,
			string variableName,
			string inlineValue)
		{
			var strings = inlineValue.Split(',').ToArray();
			var parameters = strings.SelectMany(p => p.Trim().Split(",".ToCharArray(), StringSplitOptions.RemoveEmptyEntries))
									.Select(kv => kv.Split('='))
									.Select(t => new Tuple<string, string>(t.First(), string.Join("=", t.Skip(1))))
									.Select(t => SyntaxFactory.BinaryExpression(SyntaxKind.AssignExpression, SyntaxFactory.IdentifierName(t.Item1), SyntaxFactory.IdentifierName(t.Item2.Replace("'", "\""))));

			var separators = Enumerable.Repeat(SyntaxFactory.Token(SyntaxKind.CommaToken), strings.Length - 1);
			var binding = SyntaxFactory.ObjectCreationExpression(
				SyntaxFactory.ParseTypeName("Binding"),
				null,
				SyntaxFactory.InitializerExpression(SyntaxKind.ObjectInitializerExpression, SyntaxFactory.SeparatedList<ExpressionSyntax>(parameters, separators)));
			var memberAccess = SyntaxFactory.MemberAccessExpression(SyntaxKind.MemberAccessExpression, SyntaxFactory.IdentifierName(nodeVariableName), SyntaxFactory.IdentifierName("SetBinding"));
			var dependencyProperty =
				SyntaxFactory.Argument(SyntaxFactory.MemberAccessExpression(SyntaxKind.MemberAccessExpression, SyntaxFactory.IdentifierName(nodeClassName), SyntaxFactory.IdentifierName(variableName + "Property")));
			var arguments = new[] { dependencyProperty, SyntaxFactory.Argument(binding) };
			var memberInvocation = SyntaxFactory.InvocationExpression(memberAccess, SyntaxFactory.ArgumentList(SyntaxFactory.SeparatedList(arguments, new[] { SyntaxFactory.Token(SyntaxKind.CommaToken) })));
			return SyntaxFactory.ExpressionStatement(memberInvocation);
		}

		private ExpressionSyntax GetPropertyIdentifier(XamlPropertyNode prop)
		{
			if (prop.Value == null)
			{
				return SyntaxFactory.IdentifierName(SyntaxFactory.Token(SyntaxKind.NullKeyword));
			}

			var node = prop.Value as XamlNode;
			return SyntaxFactory.IdentifierName(node != null
											 ? node.VariableName
											 : prop.Value.ToString());
		}

		private IEnumerable<StatementSyntax> WriteProperties(Func<XamlPropertyNode, IEnumerable<StatementSyntax>> propertyCreation, XamlNode childNode)
		{
			foreach (var prop in childNode.Properties)
			{
				foreach (var created in propertyCreation(prop))
				{
					yield return created;
				}

				var memberAccess = SyntaxFactory.MemberAccessExpression(SyntaxKind.MemberAccessExpression, SyntaxFactory.IdentifierName(childNode.VariableName), SyntaxFactory.IdentifierName(prop.Name));
				var setAccess = SyntaxFactory.ExpressionStatement(SyntaxFactory.BinaryExpression(SyntaxKind.AssignExpression, memberAccess, SyntaxFactory.Token(SyntaxKind.EqualsToken), GetPropertyIdentifier(prop)));

				yield return setAccess;
			}
		}

		private IEnumerable<StatementSyntax> WriteAttributes(XamlNode childNode)
		{
			foreach (var attr in childNode.Attributes)
			{
				if (!string.IsNullOrWhiteSpace(attr.ExtraCode))
				{
					yield return SyntaxFactory.ExpressionStatement(SyntaxFactory.ParseExpression(attr.ExtraCode));
				}

				var bindingMatch = BindingRegex.Match(attr.Value);
				if (bindingMatch.Success)
				{
					var inlineValue = bindingMatch.Groups[1].Value.Trim();
					yield return CreateBindingSyntax(childNode.ClassName, childNode.VariableName, attr.VariableName, inlineValue);
				}
				else if (attr.VariableName.Contains("."))
				{
					yield return CreateDependencyPropertySyntax(childNode.VariableName, attr);
				}
				else
				{
					var memberAccess = SyntaxFactory.MemberAccessExpression(SyntaxKind.MemberAccessExpression, SyntaxFactory.IdentifierName(childNode.VariableName), SyntaxFactory.IdentifierName(attr.VariableName));
					yield return SyntaxFactory.ExpressionStatement(
						SyntaxFactory.BinaryExpression(
							SyntaxKind.AssignExpression, memberAccess, SyntaxFactory.Token(SyntaxKind.EqualsToken), SyntaxFactory.IdentifierName(attr.Value)));
				}
			}
		}

		private IEnumerable<StatementSyntax> GenerateChildStatement(XamlNode childNode, Func<XamlPropertyNode, IEnumerable<StatementSyntax>> propertyCreation)
		{
			yield return CreateLocalDeclaration(childNode);
			foreach (var property in WriteProperties(propertyCreation, childNode))
			{
				yield return property;
			}

			foreach (var attribute in WriteAttributes(childNode))
			{
				yield return attribute;
			}

			if (childNode.Parent != null && !IsPropertyOf(childNode.Parent, childNode))
			{
				yield return CreateParentSyntax(childNode);
				foreach (var child in childNode.Children.SelectMany(c => GenerateChildStatement(c, propertyCreation)))
				{
					yield return child;
				}
			}
		}

		private IEnumerable<ExpressionStatementSyntax> CreatePropertyAssignment(XamlPropertyNode prop, ExpressionSyntax ownerExpression)
		{
			var values = prop.Value as IEnumerable<XamlNode>;
			if (values != null)
			{
				var statements = values.Select(v =>
					{
						var memberAccess = SyntaxFactory.MemberAccessExpression(SyntaxKind.MemberAccessExpression, ownerExpression, SyntaxFactory.IdentifierName(prop.Name));
						var addAccess = SyntaxFactory.MemberAccessExpression(SyntaxKind.MemberAccessExpression, memberAccess, SyntaxFactory.IdentifierName("Add"));
						return
							SyntaxFactory.ExpressionStatement(
								SyntaxFactory.InvocationExpression(
									addAccess,
									SyntaxFactory.ArgumentList(new[]
															   {
																   SyntaxFactory.SeparatedList(SyntaxFactory.Argument(SyntaxFactory.IdentifierName(v.VariableName)))
															   })));
					});

				foreach (var statement in statements)
				{
					yield return statement;
				}
			}
			else
			{
				var memberAccess = SyntaxFactory.MemberAccessExpression(SyntaxKind.MemberAccessExpression, ownerExpression, SyntaxFactory.IdentifierName(prop.Name));
				yield return SyntaxFactory.ExpressionStatement(SyntaxFactory.BinaryExpression(SyntaxKind.AssignExpression, memberAccess, SyntaxFactory.Token(SyntaxKind.EqualsToken), GetPropertyIdentifier(prop)));
			}
		}

		private IEnumerable<StatementSyntax> CreatePropertySyntax(XamlPropertyNode property)
		{
			var node = property.Value as XamlNode;
			if (node != null)
			{
				return GenerateChildStatement(node, CreatePropertySyntax);
			}

			var nodes = property.Value as IEnumerable<XamlNode>;
			if (nodes != null)
			{
				return nodes.SelectMany(n => GenerateChildStatement(n, CreatePropertySyntax));
			}

			return Enumerable.Empty<StatementSyntax>();
		}

		private IEnumerable<StatementSyntax> CreateSyntax(IEnumerable<XamlNode> nodes)
		{
			return nodes.SelectMany(n => GenerateChildStatement(n, CreatePropertySyntax))
					 .Select(s => s.WithTrailingTrivia(SyntaxFactory.ElasticCarriageReturnLineFeed));
		}
	}
}
